src/vmspawn/vmspawn.c

```diff
@@ -256,10 +439,38 @@ static int run_virtual_machine(void) {
                 "-m", mem,
                 "-object", "rng-random,filename=/dev/urandom,id=rng0",
                 "-device", "virtio-rng-pci,rng=rng0,id=rng-device0",
-                "-nic", "user,model=virtio-net-pci",
-                "-cpu", "max"
+                "-nic", "user,model=virtio-net-pci"
         );
 
+        bool use_vsock = arg_qemu_vsock > 0;
+        if (arg_qemu_vsock < 0) {
+                r = qemu_check_vsock_support();
+                if (r < 0)
+                        return log_error_errno(r, "Failed to check for VSock support: %m");
+
+                use_vsock = r;
+        }
+
+        unsigned int child_cid = VMADDR_CID_ANY;
+        if (use_vsock) {
+                if (arg_vsock_cid <= UINT_MAX)
+                        child_cid = (unsigned int)arg_vsock_cid;
+                else {
+                        r = machine_cid(&child_cid);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to allocate a CID for the guest: %m");
```

Comment:
hmm, this is racy. if in the meantime some other client (for example another instance of systemd-vmspawn) will do the same they might pick the same CID and collide.

isn't there some way how we can either get info from qemu about the cid it auto-assigned? Or, if we have to assign it, can't we somehow pass in a pre-allocated fd?

if nothing of this works, I think it would make sense to simply do a BSD file lock on the /dev/vhost-sock device, i.e. flock(LOCK_EX), as long as we can't be sure that qemu took possession of the CID. That would at least close the gap to clients that followe the same lock protocol, i.e. at least to all instances of vmspawn itself.

Reply:
We can pass in a vsock fd I believe, that is how mkosi has to do it because of namespace issues.

Reply:
although they also then pass in a CID alongside it, I will investigate further on this.

Reply:
I think as long as I return the file descriptor as well as the cid I can pass those both along to qemu and it should remove the race condition.

Reply:
yeah, sounds good then

Reply:
Implemented this in 7126b0eb12.